lib/de.ml

```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+          done
+          ; if len == 0 then begin
+              while !run_end - !run_start >= 11 do
+                let extra_bits = min (!run_end - !run_start - 11) 0x7F in
+                precode_freqs.(18) <- precode_freqs.(18) + 1
+                ; precode_items.(!itemptr) <- 18 lor (extra_bits lsl 5)
+                ; incr itemptr
+                ; run_start := !run_start + 11 + extra_bits
+              done
+              ; if !run_end - !run_start >= 3 then (
+                  let extra_bits = min (!run_end - !run_start - 3) 0x7 in
+                  precode_freqs.(17) <- precode_freqs.(17) + 1
+                  ; precode_items.(!itemptr) <- 17 lor (extra_bits lsl 5)
+                  ; incr itemptr
+                  ; run_start := !run_start + 3 + extra_bits)
+            end
+            else if !run_end - !run_start >= 4 then (
+              precode_freqs.(len) <- precode_freqs.(len) + 1
+              ; precode_items.(!itemptr) <- len
+              ; incr itemptr
+              ; incr run_start
+              ; while !run_end - !run_start >= 3 do
+                  let extra_bits = min (!run_end - !run_start - 3) 0x3 in
+                  precode_freqs.(16) <- precode_freqs.(16) + 1
+                  ; precode_items.(!itemptr) <- 16 lor (extra_bits lsl 5)
+                  ; incr itemptr
+                  ; run_start := !run_start + 3 + extra_bits
+                done)
+          ; while !run_start <> !run_end do
+              precode_freqs.(len) <- precode_freqs.(len) + 1
+              ; precode_items.(!itemptr) <- len
+              ; incr itemptr
+              ; incr run_start
+            done
+          ; if !run_start <> num_lens then f () else !itemptr in
+        f ()
+
+    let precompute_huffman_header c =
+      let rec f num_litlen_syms =
+        if
+          num_litlen_syms = 257
+          || c.codes.lens.litlen.(num_litlen_syms - 1) <> 0
+        then c.num_litlen_syms <- num_litlen_syms
+        else f (num_litlen_syms - 1) in
+      f num_litlen_syms
+      ; let rec g num_offset_syms =
+          if
+            num_offset_syms = 1
+            || c.codes.lens.offset.(num_offset_syms - 1) <> 0
+          then c.num_offset_syms <- num_offset_syms
+          else g (num_offset_syms - 1) in
+        g num_offset_syms
+        ; if c.num_litlen_syms <> num_litlen_syms then (
+            let max1 =
+              min (c.num_litlen_syms + c.num_offset_syms) num_litlen_syms
+              - c.num_litlen_syms in
+            let max2 = c.num_offset_syms - max1 in
+            for i = 0 to max1 - 1 do
+              c.codes.lens.litlen.(c.num_litlen_syms + i) <-
+                c.codes.lens.offset.(i)
+            done
+            ; for i = 0 to max2 - 1 do
+                c.codes.lens.offset.(i) <- c.codes.lens.offset.(max1 + i)
+              done)
+        ; c.num_precode_items <-
+            compute_precode_items
+              (Array.append c.codes.lens.litlen c.codes.lens.offset)
```

Comment:
It should be a bottle-neck but a dragoon appears here where `Array.append` has a real cost (allocation & copy).

Reply:
I had forgotten this one:
In the C codebase, the two arrays are next to each other inside a `struct`, therefore, the casting has no cost and allows to easily go trough both of them at the same time.
I had chosen to simply copy them in the first time, but it will need a thorough cleaning of the entire functions

Reply:
Ok, let's keep as is this part when it does not concern the hot-loop (compression / encoding) and focus on the hot-loop to really optimize the deflation :+1:.